x/stake/keeper.go

```diff
@@ -355,141 +388,238 @@ func (k Keeper) clearAccUpdateValidators(ctx sdk.Context) {
-		var bond DelegatorBond
-		err := k.cdc.UnmarshalBinary(bondBytes, &bond)
-		if err != nil {
-			panic(err)
-		}
-		bonds[i] = bond
+		var delegation Delegation
+		k.cdc.MustUnmarshalBinary(bondBytes, &delegation)
+		delegations = append(delegations, delegation)
 		iterator.Next()
 	}
-	return bonds[:i] // trim
+	return delegations[:i] // trim
 }
 
 // load all bonds of a delegator
-func (k Keeper) GetDelegatorBonds(ctx sdk.Context, delegator sdk.Address, maxRetrieve int16) (bonds []DelegatorBond) {
+func (k Keeper) GetDelegations(ctx sdk.Context, delegator sdk.Address, maxRetrieve int16) (bonds []Delegation) {
 	store := ctx.KVStore(k.storeKey)
-	delegatorPrefixKey := GetDelegatorBondsKey(delegator, k.cdc)
-	iterator := store.Iterator(subspace(delegatorPrefixKey)) //smallest to largest
+	delegatorPrefixKey := GetDelegationsKey(delegator, k.cdc)
+	iterator := store.SubspaceIterator(delegatorPrefixKey) //smallest to largest
 
-	bonds = make([]DelegatorBond, maxRetrieve)
+	bonds = make([]Delegation, maxRetrieve)
 	i := 0
 	for ; ; i++ {
 		if !iterator.Valid() || i > int(maxRetrieve-1) {
 			iterator.Close()
 			break
 		}
 		bondBytes := iterator.Value()
-		var bond DelegatorBond
-		err := k.cdc.UnmarshalBinary(bondBytes, &bond)
-		if err != nil {
-			panic(err)
-		}
+		var bond Delegation
+		k.cdc.MustUnmarshalBinary(bondBytes, &bond)
 		bonds[i] = bond
 		iterator.Next()
 	}
 	return bonds[:i] // trim
 }
 
-func (k Keeper) setDelegatorBond(ctx sdk.Context, bond DelegatorBond) {
+func (k Keeper) setDelegation(ctx sdk.Context, bond Delegation) {
 	store := ctx.KVStore(k.storeKey)
-	b, err := k.cdc.MarshalBinary(bond)
-	if err != nil {
-		panic(err)
-	}
-	store.Set(GetDelegatorBondKey(bond.DelegatorAddr, bond.CandidateAddr, k.cdc), b)
+	b := k.cdc.MustMarshalBinary(bond)
+	store.Set(GetDelegationKey(bond.DelegatorAddr, bond.ValidatorAddr, k.cdc), b)
 }
 
-func (k Keeper) removeDelegatorBond(ctx sdk.Context, bond DelegatorBond) {
+func (k Keeper) removeDelegation(ctx sdk.Context, bond Delegation) {
 	store := ctx.KVStore(k.storeKey)
-	store.Delete(GetDelegatorBondKey(bond.DelegatorAddr, bond.CandidateAddr, k.cdc))
+	store.Delete(GetDelegationKey(bond.DelegatorAddr, bond.ValidatorAddr, k.cdc))
 }
 
 //_______________________________________________________________________
 
 // load/save the global staking params
-func (k Keeper) GetParams(ctx sdk.Context) (params Params) {
-	// check if cached before anything
-	if !k.params.equal(Params{}) {
-		return k.params
-	}
+func (k Keeper) GetParams(ctx sdk.Context) Params {
 	store := ctx.KVStore(k.storeKey)
+	return k.getParams(store)
+}
+func (k Keeper) getParams(store sdk.KVStore) (params Params) {
 	b := store.Get(ParamKey)
 	if b == nil {
 		panic("Stored params should not have been nil")
 	}
 
-	err := k.cdc.UnmarshalBinary(b, &params)
-	if err != nil {
-		panic(err)
-	}
+	k.cdc.MustUnmarshalBinary(b, &params)
 	return
 }
+
+func (k Keeper) setNewParams(ctx sdk.Context, params Params) {
```

Comment:
Why do we have a version of `setParams` without the recalcalation check?

Reply:
I didn't want to add a comment cause I thought was self-evident :P - adding a comment in now - Need a distinct function because setParams depends on an existing previous record of params to exist (to check if maxValidators has changed) - and we panic on retrieval if it doesn't exist - hence if we use setParams for the very first params set it will panic.

Reply:
Oh OK, `setParams` could check whether a previous params exists or not but I guess this is fine too.